---
title: "BMEG 400E: Assignment 1"
output:
  html_document:
    toc: true
    toc_depth: 4
    toc_float: true
    code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Assignment Overview

For this assignment we will be using data from a Chromatin Immunoprecipitation sequencing (ChIP-seq) produced by Choi, Won-Young, et al., (2020). *https://pubmed.ncbi.nlm.nih.gov/33293480/*. This data is part of a series of experiments that aimed to understand the chromatin changes that happen when an induced Pluripotent Stem Cell (iPSC) undergo differentiation towards a Neuronal Progenitor Cell (NPC). The **fastqc** files from this experiment have been pre-downloaded and added to the server, under the following path: **/usr/local/share/data/assignment_1/**. The datasets that we will be using for the assignment are in a shared *reading only* location that you can read but ARE NOT allowed to alter.

This assignment has 3 main goals: 

  1. Get familiar with the server
  
  2. Manage a conda environment
  
  3. Perform an alignment against the human genome
  
  4. Cleaning your folder and uploading your assignment to Github


## 1. Getting Familiar with the Server

For our course we will be using a server based on a Centos system. This will be the place where you are gonna do the on-class practices and the assignments. There are a couple of things that you need to be aware while working on the server. The server has limited storage and computer power. Therefore, you will be required to be mindful of the processes you run as well as the files you keep. 

To join the server you will need to be on an active connection of a UBC Virtual Private Network (VPN). If you do not have one already, you can check how to install it here: *https://it.ubc.ca/services/email-voice-internet/myvpn/setup-documents*. Once the VPN has been set, you will need to open a terminal. 

  - **Windows system:** 
      a. Install a terminal emulator like Putty (*https://www.chiark.greenend.org.uk/~sgtatham/putty/latest.html*). 
      b. This will allow you to do a SSH connection using the server's IP (137.82.55.186) and your credentials. 
  
  - **Linux/Unix system (Apple computer and Ubuntu):** 
      a. Open a terminal
      b. Type the following command: ssh username@137.82.55.186 
      c. When prompted, type your password to complete the login 
      
    

Now that you have successfully joined the server, let's get you used to some basic commands

### a. Creating a directory 

```{bash, eval=FALSE}

## Create a new directory (folder) for this assignment - 1pt 
## Tip: look at the help of the *mkdir* command (man mkdir)
mkdir Assignment_1

```

*Note:* Generally speaking, is good to follow naming convention when using the terminal. Remember: 

  - Do not start a name with a number

  - Names are case sensitive (ASSIGNMENT1.txt and assignment1.txt are not the same)

  - Avoid to use spaces, as they are interpreted as special characters by the computer

  - Use _ or - to replace spaces

  - File extensions are meaningful for us to know the file format but in terminal you can use it as part of the file name. Ex: you will be able to open a .sequences file that has tab delimited information
  
  
### b. Check a directory's path 

When using terminal, paths work as the addresses of the files and directories you want to access and easily move between them.

```{bash, eval=FALSE}
## Check your current directory's path: 0.5 pt
## Tip: look at the documentation of the *pwd* command (man pwd)
pwd

```

### c. Moving within directories

Access the newly created directory

```{bash, eval=FALSE}
## Move to your newly created directory - 0.5 pt 
## Tip: look at the *cd* command documentation
cd assignment_1

```


How would you move back to your home directory (*/home/your_user_name*)?

```{bash, eval=FALSE}

#Check the tutorial: https://www.computerhope.com/unix/ucd.htm 

## Using the complete directory path of your home directory: - 0.25 pt
cd /home/your_username/

## Using the "go to parent directory" shortcut - 0.25
cd .. 


```


### d. Explore the dataset 

The sequencing data that we will be using is paired-end. This means that each sequence has been sequenced twice, one on each end (5' and 3'). Choose one of the reads files (1 or 2) for the following exercises.


```{bash, eval=FALSE}

## Look at the first 5 rows of the dataset - 0.25
# Tip: look at the *head* command documentation (man head)
tail -5  /usr/local/share/data/assignment_1/SRR12506919_1_subset.fastq

## Look at the last 6 rows of the dataset - 0.25
# Tip: look at the *tail* command documentation (man tail)
tail -6  /usr/local/share/data/assignment_1/SRR12506919_1_subset.fastq

## Explore the file with the *less* command (Exit pressing: q) - 0.25 
less /usr/local/share/data/assignment_1/SRR12506919_1_subset.fastq


```

### e. Piping 

Because this is a very large dataset we will proceed to subset it in order to make it more manageable for the assignment. Using the commands that you learned above:

```{bash, eval=FALSE}

## How many lines does the file was in total? - 0.25
#Tip: loop at the *wc* command documentation (man wc)
wc -l SRR12506919_2_subset.fastq


## Select only the id lines (e.g. @SRR12506919.667552 667552 length=151) of the dataset (the ones that start with @ and are followed by the read id) - 0.75 pt
## Tip: look at the *grep* command
grep "@" SRR12506919_2_subset.fastq

## How many reads are in the file? - 1 pt
## Tip: Try using * | head * after the command line you use for the previous question
grep "@" SRR12506919_2_subset.fastq | head
grep "@" SRR12506919_2_subset.fastq | wc -l  ## 667557


## Select only the reads id (e.g. @SRR12506919.667552) from the id lines - 1 pt
## Tip: Look into the *cut* command. Carefully read the default delimiter, is it the case for our file?
grep "@" SRR12506919_2_subset.fastq | cut -f1 --delimiter=" " 


```


### f. Saving an output 


```{bash, eval=FALSE}

## Save a file that contains only the reads ids (the result of our previous exercise). - 0.5 pt
## Tip: Search in Google how to save an output on bash/terminal 
grep "@" SRR12506919_2_subset.fastq | cut -f1 --delimiter=" "  > SRR12506919reads_ids.txt

## Now, list all the files in a directory: 
## Tip: look at the *ls* command

# What do you see? Was the subset file created correctly? - 0.25 pt
## ----

```


### g. Creating a backup 

There will be times where you will want to save a copy of a dataset or file before modifying in case something goes wrong. For those cases, you can create a copy of any file or directory using the "copy" command

```{bash, eval=FALSE}

## Create a copy of the reads ids file - 0.25 pt
## Tip: man cp 
cp SRR12506919reads_ids.txt SRR12506919reads_ids_copy.txt

## Change the name of the backup reads id file - 0.25 pt
## Tip: man mv
mv SRR12506919reads_ids_copy.txt new_name

## Delete the copy of the reads id file - 0.25 pt
## Tip: man rm
rm new_name

```

## 2. Managing Conda Environments

### a. Create a conda environment

In order to run the reads alignments against the human genome, there are a few tools that we will need:

  - fastQC (*https://www.bioinformatics.babraham.ac.uk/projects/fastqc/*): comprehensive quality control measures of sequencing data.
  
  
  - bowtie2 (*http://bowtie-bio.sourceforge.net/bowtie2/index.shtml*): alignments of sequencing data. 
  
  - samtools (*http://www.htslib.org/doc/samtools.html*): set of tools to manage sam/bam alignment files 
  
  
To install them, we will be making use of the conda environments. Conda allows you to create and manage environments for any programming language. Managing this environments mean that you can work with specific versions of different programs at any given moment, simply by loading the desired environment. You can find more information about this resource here: *https://docs.conda.io/en/latest/* . 

```{bash, eval=FALSE}
## Create a new conda environment: - 0.5 pt
# Tip: Consult the previously provide links or consult the conda create help (conda create --help)

conda create --name BMEDG_1

```


### b. Add programs to your conda environment

Now that the environment has been created, its time to add the packages that we will need. Conda has an active community and a great documentation (*https://docs.conda.io/projects/conda/en/latest/user-guide/tasks/manage-environments.html*) which you can use throughout the course to help answer any questions you may have. 


```{bash, eval=FALSE}

## Add fastQC bowtie2 to your conda environment: - 1 pt
conda install -c bioconda -n BMEDG_1 fastqc bowtie2 

## Run the following command to install samtools onto your conda environment:
conda install -n your_environment_name  samtools==1.11

```


## 3. Performing Alignments 

### a. Data quality check

We will use the widely used fastQC software to do a quick inspection of the data quality. Once it has ran, it will give you an html report of the quality of the data, that you can open using a web browser. More information on how to read the output can be found here: *https://dnacore.missouri.edu/PDF/FastQC_Manual.pdf* and in the tool's website.

  
```{bash, eval=FALSE}

## IMPORTANT NOTE: Remember to activate your conda environment!

## Run fastQC on the fastq files: - 1pt
fastqc usr/local/share/data/assignment_1/SRR12506919_1_subset.fastq usr/local/share/data/assignment_1/SRR12506919_2_subset.fastq

## In order to open the html files on your web browser you will need to download the files to your computer
## Ubuntu/Linux and Mac users: look at the *scp* command
## Windows users: follow the following instructions: https://stackoverflow.com/questions/6217055/how-can-i-copy-a-file-from-a-remote-server-to-using-putty-in-windows

## What can you say about the data quality? - 2 pt 
# ---

```
  
  
### b. Running a process on the background: screen 

The processes that we are about to run, can take a long time to finish. Thus, we will make use of the *screen* tool, which allows you to run a process in the background while continue using your server session without the risk of ending the process due to a bad internet connection, accidentally closing the tab or other random circumstances. 

```{bash, eval=FALSE}

## To run a process in a background screen with screen you:

# 1. Start a background process with a specific name 
screen -S background_screen_name 
# 2. Run the process and any commands you wish
wc -l usr/local/share/data/assignment_1/SRR12506919_1_subset.fastq

# 3. Get out of the background screen, you will need to type the following:
ctrl + A + D 

# 4. Return to the background screen to check the process
screen -r background_screen_name

# 5. Terminate the background screen once the process has ended
# Within the background screen type:
exit 

```


### c. Checking the server capacity and your current runs: htop

Another way to check on the progress of your runs and the server status is with the *htop* command. This will open a screen showing all the processes that are being currently being run in the server. Our server only has 2 cpu's/cores, the green bar next to each code, represents how much of that node it is currently in use. Always make sure to check the processes that are being run before sending yours.

```{bash, eval=FALSE}

## Use the htop command to describe the status of the server - 1.5 pt


```



### d. Genome indexing - bowtie2

Now, we will need to create an index of the human genome that bowtie2 will use to map our sequences in the genome. In order to do this, you will need to use the previously downloaded files of the human genome with the desired build (e.g. hg19,hg38), you can find those files within the server here: */usr/local/share/human/*

**BEFORE RUNNING ANYTHING**: go to the "Other resources" section at the end of the assignment!


```{bash, eval=FALSE}

## Something useful to do when using a new software is to look at the documentation using the *help* option
## Try running: 
bowtie2 -h 

## IMPORTANT!!!!
## BEFORE RUNNING: go to the "Other resources" section at the end of the assignment!
## __________________________________________________________________________________
## Use the hg38 build to create an index of the human genome with bowtie2
## Tip: look into the bowtie2-build help (bowtie2-build --help)  - 1.5 pt

== 17:18:21
bowtie2-build /usr/local/share/human/hg38/hg38.fa DELETE_index
== 


```

### e. Alignment

We are working with paired-end data. Thus, you will need to make sure to use both fastq files to align the sequences to the genome.
**IMPORTANT:** Run with *default parameters* DO NOT specify any non-essential paramaters.

**Time flag**: This step will take up to 30 mins 

```{bash, eval=FALSE}

## Perform a paired-end alignment of the fastq sequences provided (located here: ) to the human genome index build in the previous step - 2pt
## Tip: look at the bowtie2 --help or the documentation on the tool website (http://bowtie-bio.sourceforge.net/bowtie2/manual.shtml#getting-started-with-bowtie-2-lambda-phage-example)


== 17:16:39
bowtie2 -x hg38_bowtie2_index \
-1 /usr/local/share/data/assignment_1/SRR12506919_1_subset.fastq \
-2 /usr/local/share/data/assignment_1/SRR12506919_2_subset.fastq \
-S SRR12506919_paired_end_alignment.sam
== 17:42:48

```


### f. Viewing the alignments

Now, we will make use of **samtools** to review some basic features of the alignment. For mor information about this tool: *http://www.htslib.org/doc/*

```{bash, eval=FALSE}

## Use *samtools view* to get the:
## Tip: check out the different flag (-f and -F) options
## Tip: Read the samtools view --help, read carefully for an option that allows you to *count* the results of your search

## Number of mapped reads - 1 pt
samtools view -c -F 4 SRR12506919_paired_end_alignment.sam 
## 1223624 reads

## Mumber of unmapped reads - 1 pt
samtools view -c -f 4 SRR12506919_paired_end_alignment.sam 
## 111490 reads


```


## 4. Cleaning and Uploading

### a. Cleaning your folders

Before signing up, we need to make sure that we won't leave behind any big files that can take up a lot of memory from our server. To do this, make sure to:

  a. Delete any copies of the input assignment files you might have done on your personal folder
  b. Zip or delete all the files used for the assignment 


### b. Uploading to Github

ADD ON HOW TO DOWNLOAD AN RMD AND UPLOAD IT TO THEIR GITHUB REPOSITORIES
Save and upload your final report onto your github repository!

## Other Resources

### a. Bowtie2 index

Indexes take a lot of computational resources and time to run. The one we need for the human genome will take around 1.5 hours to be done. ***DO NOT RUN THE INDEX COMMAND***. Go on to the next step, using the previously run index: *hg38_bowtie2_index* under the following path: */usr/local/share/indexes/*

